### ISSUE TITLE
Kubernetes 1.2.4 - Incorrectly reporting KubeletHasSufficientDisk

### ISSUE THREAD

**Issue by ajohnstone** [kubernetes/kubernetes#26566]
The kubernetes node has run out of space, however the cluster is reporting as having available disk space. The LastHeartbeatTime is up to date.

For full details: https://gist.github.com/ajohnstone/00d3e9da2b4329302d8972fa7419cc9e

```
Conditions:
  Type      Status  LastHeartbeatTime           LastTransitionTime          Reason              Message
  ----      ------  -----------------           ------------------          ------              -------

$ kubectl describe nodes  2>&1 | grep OutOfDisk
  OutOfDisk     False   Tue, 31 May 2016 14:28:48 +0100     Sat, 28 May 2016 16:00:13 +0100     KubeletHasSufficientDisk    kubelet has sufficient disk space available
  OutOfDisk     False   Tue, 31 May 2016 14:28:43 +0100     Mon, 30 May 2016 23:43:18 +0100     KubeletHasSufficientDisk    kubelet has sufficient disk space available
  OutOfDisk     False   Tue, 31 May 2016 14:28:42 +0100     Mon, 16 May 2016 13:41:48 +0100     KubeletHasSufficientDisk    kubelet has sufficient disk space available
  OutOfDisk     False   Tue, 31 May 2016 14:28:46 +0100     Sat, 07 May 2016 23:20:21 +0100     KubeletHasSufficientDisk    kubelet has sufficient disk space available
  OutOfDisk     False   Tue, 31 May 2016 14:28:48 +0100     Mon, 30 May 2016 09:23:19 +0100     KubeletHasSufficientDisk    kubelet has sufficient disk space available
```

The nodes filesystem usage, on /mnt/ephemeral is 100%.

```
ip-10-115-0-117 $ df -h
Filesystem                           Size  Used Avail Use% Mounted on
/dev/xvda1                           493G  2.4G  470G   1% /
udev                                  10M     0   10M   0% /dev
tmpfs                                3.0G  297M  2.7G  10% /run
tmpfs                                7.4G  1.5M  7.4G   1% /dev/shm
tmpfs                                5.0M     0  5.0M   0% /run/lock
tmpfs                                7.4G     0  7.4G   0% /sys/fs/cgroup
/dev/mapper/vg--ephemeral-ephemeral   74G   70G  238M 100% /mnt/ephemeral
100% /mnt/ephemeral
tmpfs                                7.4G   12K  7.4G   1% /mnt/ephemeral/kubernetes/kubelet/pods/b9288c4f-14a1-11e6-bf17-0a400f658983/volumes/kubernetes.io~secret/default-token-mibkw
tmpfs                                7.4G   12K  7.4G   1% /mnt/ephemeral/kubernetes/kubelet/pods/9c843917-14a3-11e6-bf17-0a400f658983/volumes/kubernetes.io~configmap/nginx-template-volume
tmpfs                                7.4G   12K  7.4G   1% /mnt/ephemeral/kubernetes/kubelet/pods/7bea4847-1f04-11e6-bf17-0a400f658983/volumes/kubernetes.io~secret/default-token-5z4dc
tmpfs                                7.4G   12K  7.4G   1% /mnt/ephemeral/kubernetes/kubelet/pods/a7a0571c-1dd3-11e6-bf17-0a400f658983/volumes/kubernetes.io~secret/default-token-gyh7f
tmpfs                                7.4G  4.0K  7.4G   1% /mnt/ephemeral/kubernetes/kubelet/pods/9c843917-14a3-11e6-bf17-0a400f658983/volumes/kubernetes.io~configmap/nginx-confd-volume
tmpfs                                7.4G   12K  7.4G   1% /mnt/ephemeral/kubernetes/kubelet/pods/9c843917-14a3-11e6-bf17-0a400f658983/volumes/kubernetes.io~secret/default-token-5z4dc
tmpfs                                7.4G   12K  7.4G   1% /mnt/ephemeral/kubernetes/kubelet/pods/fa98a0e0-20f9-11e6-bf17-0a400f658983/volumes/kubernetes.io~secret/default-token-5z4dc
tmpfs                                7.4G   12K  7.4G   1% /mnt/ephemeral/kubernetes/kubelet/pods/54b54562-2285-11e6-bf17-0a400f658983/volumes/kubernetes.io~secret/default-token-gyh7f
tmpfs                                7.4G   12K  7.4G   1% /mnt/ephemeral/kubernetes/kubelet/pods/7ce74ae0-228c-11e6-bf17-0a400f658983/volumes/kubernetes.io~secret/default-token-qgr6z
tmpfs                                7.4G   12K  7.4G   1% /mnt/ephemeral/kubernetes/kubelet/pods/0f181767-2357-11e6-bf17-0a400f658983/volumes/kubernetes.io~secret/default-token-qgr6z
tmpfs                                7.4G   12K  7.4G   1% /mnt/ephemeral/kubernetes/kubelet/pods/6785523b-23e5-11e6-bf17-0a400f658983/volumes/kubernetes.io~secret/default-token-8dysa
Raw
```

The api data.

```
kubectl get nodes ip-10-115-0-117.eu-west-1.compute.internal -o json |  jq 'del(.status.images)' 
{
  "kind": "Node",
  "apiVersion": "v1",
  "metadata": {
    "name": "ip-10-115-0-117.eu-west-1.compute.internal",
    "selfLink": "/api/v1/nodes/ip-10-115-0-117.eu-west-1.compute.internal",
    "uid": "e31b5033-14a1-11e6-bf17-0a400f658983",
    "resourceVersion": "1258293",
    "creationTimestamp": "2016-05-07T22:20:21Z",
    "labels": {
      "beta.kubernetes.io/instance-type": "m3.xlarge",
      "failure-domain.beta.kubernetes.io/region": "eu-west-1",
      "failure-domain.beta.kubernetes.io/zone": "eu-west-1c",
      "kubernetes.io/hostname": "ip-10-115-0-117.eu-west-1.compute.internal"
    }
  },
  "spec": {
    "podCIDR": "100.66.2.0/24",
    "externalID": "i-1b922897",
    "providerID": "aws:///eu-west-1c/i-1b922897"
  },
  "status": {
    "capacity": {
      "cpu": "4",
      "memory": "15434188Ki",
      "pods": "110"
    },
    "allocatable": {
      "cpu": "4",
      "memory": "15434188Ki",
      "pods": "110"
    },
    "conditions": [
      {
        "type": "OutOfDisk",
        "status": "False",
        "lastHeartbeatTime": "2016-05-31T13:22:32Z",
        "lastTransitionTime": "2016-05-30T22:43:18Z",
        "reason": "KubeletHasSufficientDisk",
        "message": "kubelet has sufficient disk space available"
      },
      {
        "type": "Ready",
        "status": "True",
        "lastHeartbeatTime": "2016-05-31T13:22:32Z",
        "lastTransitionTime": "2016-05-30T22:43:18Z",
        "reason": "KubeletReady",
        "message": "kubelet is posting ready status. WARNING: CPU hardcapping unsupported"
      }
    ],
    "addresses": [
      {
        "type": "InternalIP",
        "address": "10.115.0.117"
      },
      {
        "type": "LegacyHostIP",
        "address": "10.115.0.117"
      },
      {
        "type": "ExternalIP",
        "address": "54.171.148.156"
      }
    ],
    "daemonEndpoints": {
      "kubeletEndpoint": {
        "Port": 10250
      }
    },
    "nodeInfo": {
      "machineID": "1d7ef93259a343c2a20e96407fcb7444",
      "systemUUID": "EC26E0C6-8C3C-0634-A69B-DE28D91118CC",
      "bootID": "0d6d130b-c8b6-4ae4-95df-9c237d84fa19",
      "kernelVersion": "3.16.0-4-amd64",
      "osImage": "Debian GNU/Linux 8 (jessie)",
      "containerRuntimeVersion": "docker://1.9.1",
      "kubeletVersion": "v1.2.4",
      "kubeProxyVersion": "v1.2.4"
    }
  }
}
```

**Comment by a-robinson** [comment]
@kubernetes/sig-node

**Comment by vishh** [comment]
Kubelet is looking at `/dev/xvda1` which is mounted at `/` and deciding
that the node has enough disk space. It is not detecting that docker daemon
is made to use `/dev/mapper/vg--ephemeral-ephemeral` which is mounted at
`/mnt/ephemeral`.

This might be a kubelet (or a cAdvisor) bug.

**Comment by ncdc** [comment]
@ajohnstone can you gist/paste the output of `sudo docker info` please? Thanks!

**Comment by ncdc** [comment]
@vishh I'll need the `docker info` data to confirm, but it looks like there are 3 possible devices in play here:

1. imagefs
1. rootfs
1. device on which the kubelet's volume directory resides

Looking at the output above, it looks like the kubelet's volume directory is not part of rootfs, but is instead on `/dev/mapper/vg--ephemeral-ephemeral`. I'm thinking that imagefs and rootfs are not `OutOfDisk` but the device for the volume directory is.

Maybe we need the kubelet to account for the volume directory too?

**Comment by ajohnstone** [comment]
This deployment is just using the kube-up.sh with aws.
```
Containers: 18
Images: 87
Server Version: 1.9.1
Storage Driver: aufs
 Root Dir: /mnt/ephemeral/docker/aufs
 Backing Filesystem: extfs
 Dirs: 123
 Dirperm1 Supported: true
Execution Driver: native-0.2
Logging Driver: json-file
Kernel Version: 3.16.0-4-amd64
Operating System: Debian GNU/Linux 8 (jessie)
CPUs: 4
Total Memory: 14.72 GiB
Name: ip-10-115-0-126
ID: VVGV:HE6H:OAIE:UR7G:Z3P5:CWAT:5PRA:EU3I:V6TW:TGQM:TFVN:2R2Y
WARNING: No swap limit support
```

**Comment by ncdc** [comment]
@ajohnstone ok, so in this case it should have detected that imagefs was OutOfDisk, I believe. I wonder why it didn't...

**Comment by ronnielai** [comment]
The isSpaceAvailable check is done by comparing the available with the threshold. /mnt/ephemeral still has 238M available. Is it possible that the available is actually larger than the threshold?

**Comment by vishh** [comment]
@ncdc I suspect that cAdvisor is incorrectly discovering docker root directory and hence the docker image partition.

@ajohnstone Can you post the logs from kubelet? (`/var/log/kubelet.log`)

**Comment by timstclair** [comment]
Other helpful debugging information:
- `docker info` (specifically what does it report for the root dir)
- cAdvisor storage report: `<node host>:4194/api/v2.1/storage`

I wonder whether this is fixed by https://github.com/google/cadvisor/pull/1275, which reads the docker root directory from the docker status rather than the `--docker-root` flag.

**Comment by ajohnstone** [comment]
@vishh unfortunately I have deleted that particular cluster already. I will have to create another and reproduce the problem.

**Comment by ajohnstone** [comment]
I've found the cause of the above... 
```
Jun  6 18:06:05 ip-10-116-0-120 kubelet[5672]: I0606 18:06:05.021682    5672 conversion.go:128] failed to handle multiple devices for container. Skipping Filesystem stats
Jun  6 18:06:05 ip-10-116-0-120 kubelet[5672]: I0606 18:06:05.021712    5672 conversion.go:128] failed to handle multiple devices for container. Skipping Filesystem stats
Jun  6 18:06:05 ip-10-116-0-120 kubelet[5672]: I0606 18:06:05.031870    5672 server.go:1096] GET /stats/summary/: (17.917727ms) 200 [[Go-http-client/1.1] 10.116.0.124:47400]
```

stats
```
$ curl -s 0:4194/api/v2.1/storage | python -m json.tool;
[
    {
        "available": 504282198016,
        "capacity": 528303898624,
        "device": "/dev/xvda1",
        "labels": [
            "root"
        ],
        "mountpoint": "/",
        "usage": 2445176832
    },
    {
        "available": 6494826496,
        "capacity": 79107727360,
        "device": "/dev/mapper/vg--ephemeral-ephemeral",
        "labels": [
            "docker-images"
        ],
        "mountpoint": "/mnt/ephemeral",
        "usage": 68570853376
    }
]
```
docker info
```
$ docker info;
Containers: 29
Images: 245
Server Version: 1.9.1
Storage Driver: aufs
 Root Dir: /mnt/ephemeral/docker/aufs
 Backing Filesystem: extfs
 Dirs: 301
 Dirperm1 Supported: true
Execution Driver: native-0.2
Logging Driver: json-file
Kernel Version: 3.16.0-4-amd64
Operating System: Debian GNU/Linux 8 (jessie)
CPUs: 4
Total Memory: 14.72 GiB
Name: ip-10-116-0-120
ID: PJTZ:UYTA:Z7NI:OIGF:ABJQ:2QO4:AJZN:W3U5:5SF7:AGFO:UVDI:N6NM
WARNING: No swap limit support
```

**Comment by vishh** [comment]
This error `conversion.go:128] failed to handle multiple devices for container. Skipping Filesystem stats` should probably be suppressed. It does not affect kubelet functionalities.
Can you post the output of `http://localhost:10255/stats/summary`?

**Comment by timstclair** [comment]
That's strange, the numbers on the stats you posted don't add up. I think `"usage"` should equal `"capacity" - "available"` (computed [here](https://github.com/google/cadvisor/blob/v0.22.2/fs/fs.go#L497))

**Comment by ajohnstone** [comment]
I've run some more tests, and it does seem to be behaving correctly, at least it is now with my current cluster.

I'm a little unsure how the thresholds are set "available 120 MB, threshold 256 MB"

the full details can be found at https://gist.github.com/ajohnstone/ab8632ae4e8fc573bb66cac3ce0881f8

The log "failed to handle multiple devices for container. Skipping Filesystem stats" did confuse me.
----

I filled the disk to  +1gb free capacity.
```
$ FILL=$(df | awk '/ephemeral$/ {print int(($4/1024/1024-1)*1024)}')
$ dd if=/dev/zero of=test.dat bs=$((1024*1024)) count=$FILL;
64653+0 records in
64653+0 records out
67793584128 bytes (68 GB) copied, 432.068 s, 157 MB/s
```

Repeated tests filling disk space.
```
$ kubectl get nodes -o json ip-10-116-0-120.eu-west-1.compute.internal  | jq '.status.conditions[] | select(.type =="OutOfDisk") ' 
{
  "type": "OutOfDisk",
  "status": "False",
  "lastHeartbeatTime": "2016-06-06T19:54:28Z",
  "lastTransitionTime": "2016-06-06T19:28:50Z",
  "reason": "KubeletHasSufficientDisk",
  "message": "kubelet has sufficient disk space available"
}


$ ssh -i ~/.ssh/photobox-eu admin@54.229.153.20 sudo 'dd if=/dev/zero of=/mnt/ephemeral/docker/aufs/mnt/test.2.dat bs=1048576 count=500
500+0 records in
500+0 records out
524288000 bytes (524 MB) copied, 0.40629 s, 1.3 GB/s

$ ssh -i ~/.ssh/photobox-eu admin@54.229.153.20 df  /mnt/ephemeral
Filesystem                          1K-blocks     Used Available Use% Mounted on
/dev/mapper/vg--ephemeral-ephemeral  77253640 72772180    534148 100% /mnt/ephemeral

$ ssh -i ~/.ssh/photobox-eu admin@54.229.153.20 sudo 'dd if=/dev/zero of=/mnt/ephemeral/docker/aufs/mnt/test.2.dat bs=1048576 count=300'
300+0 records in
300+0 records out
314572800 bytes (315 MB) copied, 0.933002 s, 337 MB/s

$ ssh -i ~/.ssh/photobox-eu admin@54.229.153.20 df  /mnt/ephemeral
Filesystem                          1K-blocks     Used Available Use% Mounted on
/dev/mapper/vg--ephemeral-ephemeral  77253640 72567572    738756  99% /mnt/ephemeral

kubernetes.src master → kubectl get nodes -o json ip-10-116-0-120.eu-west-1.compute.internal  | jq '.status.conditions[] | select(.type =="OutOfDisk") ' 
{
  "type": "OutOfDisk",
  "status": "False",
  "lastHeartbeatTime": "2016-06-06T20:00:20Z",
  "lastTransitionTime": "2016-06-06T19:28:50Z",
  "reason": "KubeletHasSufficientDisk",
  "message": "kubelet has sufficient disk space available"
}

$ ssh -i ~/.ssh/photobox-eu admin@54.229.153.20 sudo 'dd if=/dev/zero of=/mnt/ephemeral/docker/aufs/mnt/test.2.dat bs=1048576 count=900'
900+0 records in
900+0 records out
943718400 bytes (944 MB) copied, 2.76576 s, 341 MB/s

$ kubectl get nodes -o json ip-10-116-0-120.eu-west-1.compute.internal  | jq '.status.conditions[] | select(.type =="OutOfDisk") ' 
{
  "type": "OutOfDisk",
  "status": "True",
  "lastHeartbeatTime": "2016-06-06T20:00:30Z",
  "lastTransitionTime": "2016-06-06T20:00:30Z",
  "reason": "KubeletOutOfDisk",
  "message": "out of disk space"
}

```

**Comment by vishh** [comment]
Having access to kubelet logs will help. Unfortunately, kubelet does not log errors on disk space checks [here](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet.go#L2467).

**Comment by ronnielai** [comment]
I think a log is generated for kubelet disk space check errors in the lower level by [isSpaceAvailable](https://github.com/kubernetes/kubernetes/blob/6224f44717effa17914fac41f13a4f2be7b5c21b/pkg/kubelet/disk_manager.go#L94).

**Comment by dohnto** [comment]
Hello, we face similar (related) situation:

Docker is not able to create new containers


```
docker run -it debian
Unable to find image 'debian:latest' locally
Trying to pull repository docker.io/library/debian ... 
latest: Pulling from docker.io/library/debian
75a822cd7888: Already exists 
Digest: sha256:f7062cf040f67f0c26ff46b3b44fe036c29468a7e69d8170f37c57f2eec1261b
Status: Downloaded newer image for docker.io/debian:latest
docker: Error response from daemon: devmapper: Thin Pool has 111339 free data blocks which is less than minimum required 147456 free data blocks. Create more free space in thin pool or use dm.min_free_space option to change behavior.
See '/usr/bin/docker-current run --help'.
```

And if pod was scheduled to such a node, it repeatedly fails and the pod is never rescheduled to other node! This can cause outage!

### docker info
```
Containers: 102
 Running: 49
 Paused: 0
 Stopped: 53
Images: 26
Server Version: 1.10.3
Storage Driver: devicemapper
 Pool Name: docker-253:0-146166401-pool
 Pool Blocksize: 65.54 kB
 Base Device Size: 107.4 GB
 Backing Filesystem: xfs
 Data file: /dev/fedora/docker-data
 Metadata file: /dev/fedora/docker-metadata
 Data Space Used: 87.86 GB
 Data Space Total: 96.64 GB
 Data Space Available: 8.777 GB
 Metadata Space Used: 94.93 MB
 Metadata Space Total: 4.295 GB
 Metadata Space Available: 4.2 GB
 Udev Sync Supported: true
 Deferred Removal Enabled: true
 Deferred Deletion Enabled: true
 Deferred Deleted Device Count: 0
 Library Version: 1.02.122 (2016-04-09)
Execution Driver: native-0.2
Logging Driver: journald
Plugins: 
 Volume: local
 Network: bridge null host
Kernel Version: 4.8.12-200.fc24.x86_64
Operating System: Fedora 24 (Atomic Host)
OSType: linux
Architecture: x86_64
Number of Docker Hooks: 2
CPUs: 24
Total Memory: 62.9 GiB
Name: vs-sklik-brno-11.dev
ID: CKQD:NN3D:EA35:YKQI:37AW:KIML:KWOD:ETSK:QM2Y:K2FA:DRJK:TQFN
Registries: docker.io (secure)
```

### kubectl describe node ournode
```
OutOfDisk 	False 	Wed, 21 Dec 2016 16:09:29 +0100 	Wed, 21 Dec 2016 11:15:24 +0100 	KubeletHasSufficientDisk 	kubelet has sufficient disk space available
```

**Comment by cmluciano** [comment]
Is this still an issue? It looks like locks on a thin pool are not counted as @dohnto stated here https://github.com/kubernetes/kubernetes/issues/26566#issuecomment-268790634. This type of issue would be specific to the devicemapper storage driver.
